File: leetcode/1475-final-prices-with-a-special-discount-in-a-shop/solution.py
"""
* Author: 1chooo<hugo970217@gmail.com>
* Problem: https://leetcode.com/problems/final-prices-with-a-special-discount-in-a-shop/
* Runtime: 3ms (54.81%)
"""

class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        stack = []

        for i, price in enumerate(prices):
            while stack and prices[stack[-1]] >= price:
                prices[stack.pop()] -= price
            stack.append(i)

        return prices

File: leetcode/1475-final-prices-with-a-special-discount-in-a-shop/solution.cpp
/**
 * Author: 1chooo<hugo970217@gmail.com>
 * Problem: https://leetcode.com/problems/final-prices-with-a-special-discount-in-a-shop/
 * Runtime: 0ms (100.00%)
 */

const static auto _ = []() {
    cin.tie(nullptr)->sync_with_stdio(false);
    return nullptr;
}();

class Solution {
public:
    vector<int> finalPrices(vector<int>& prices) {
        stack<int> s;

        for (int i = 0; i < prices.size(); i++) {
            while (!s.empty() && prices[s.top()] >= prices[i]) {
                prices[s.top()] -= prices[i];
                s.pop();
            }
            s.push(i);
        }

        return prices;
    }
};
